### ISSUE TITLE
How do I add a `kid` to the OIDC `id_token`?

### ISSUE THREAD

**Issue by jeffsawatzky** [lepture/authlib#367]
I would like to add a `kid` to the header of the OIDC `id_token` but I can't seem to figure out how to do so.
This is needed to be able to use the discovery keys endpoint which lists the public keys and their `kid`'s.

I'm using authlib 0.15.4

**Comment by jeffsawatzky** [comment]
According to section 10.1 of the openid connect spec, a kid value MUST be specified in the JOSE header if there are multiple keys, which would be the case when we rotate our keys.

https://openid.net/specs/openid-connect-core-1_0.html#SigEnc

**Comment by lepture** [comment]
@jeffsawatzky I think this is a bug in current implementation.

**Comment by lepture** [comment]
@jeffsawatzky this is already fixed in v1.0.0a2, it will pick a key from your JWKs randomly, and add the `kid` into `id_token`.

**Comment by dwt** [comment]
This may also help, I had to provide the JWT_CONFIG in all `dict` format, instead of as a `KeySet`, else the`kid` wasn't added to the generated `id_token`.

```python
JWT_CONFIG = {
    # FIXME get secret from secrets.cfg
    "key": "secret-key",
    "alg": "RS256",
    "iss": "https://sntl-publishing.com",
    "exp": 3600,
}
private_key_path = Path('etc') / 'private.pem'
private_key = JsonWebKey.import_key(private_key_path.read_text())
JWT_CONFIG['key'] = KeySet([private_key]).as_dict()
```

**Comment by lepture** [comment]
@jeffsawatzky It is not easy to fix in 0.15.x. This is already fixed in master code. Currently you can use `v1.0.0b1` release. I'll focus on releasing v1.0.0 in next month.

I'm using the beta release of v1.0.0 myself in [Typlog](https://typlog.com/). I'll close it.

**Comment by jeffsawatzky** [comment]
For those that are trying to figure out how to do this, I sort of followed what @dwt suggested, however instead of creating a `KeySet` and calling `as_dict`, I just called `as_dict` directly on the `private_key` and passed that. I ALSO had to pass `is_private=True` to `as_dict` otherwise it would generate the public key dict, which you can't use to sign with.

```
private_key_path = Path('etc') / 'private.pem'
private_key = JsonWebKey.import_key(private_key_path.read_text())

jwt_config = {
    "key": private_key.as_dict(is_private=True),
    "alg": "RS256",
    "iss": "https://example.com",
    "exp": 3600,
}
```
As mentioned above, you can't just pass the loaded key because the `kid` property will be `None`. It seems like converting to a `dict` is the only way to get the `kid` to be set, which will call `thumbprint` if it isn't set. @lepture I don't know if this is expected behaviour not, but I cold not get a `kid` set by passing an `RSAKey` or a `KeySet` with the `RSAKey` as it doesn't seem to fall back to the `thumbprint` if a `kid` isn't set in the tokens or something like that.
